Improve SmallRye Health extension's default scope for health checks

Ladicek opened the pull request:
The SmallRye Health extension registers health check annotations
as additional bean defining annotations, but doesn't define a default
scope for them. Instead, the extension used to produce an annotation
transformation that adds `@Singleton` to every class or method annotated
with one of the health check annotations and not annotated with a scope
annotation or a stereotype with a scope annotation.

This commit removes that annotation transformation and instead defines
that the health checks annotations that are registered as BDAs have
a default scope of `@Singleton`. Less code, same outcome.

Ladicek committed:
Improve SmallRye Health extension's default scope for health checks

The SmallRye Health extension registers health check annotations
as additional bean defining annotations, but doesn't define a default
scope for them. Instead, the extension used to produce an annotation
transformation that adds `@Singleton` to every class or method annotated
with one of the health check annotations and not annotated with a scope
annotation or a stereotype with a scope annotation.

This commit removes that annotation transformation and instead defines
that the health checks annotations that are registered as BDAs have
a default scope of `@Singleton`. Less code, same outcome.

mkouba commented:
Let's run the CI...

mkouba commented:
The windows CI failure is caused by two problems:
1. The tests are executed in different order (compared to linux machines), i.e. `HealthCheckContextPropagationTest` is executed before `HealthCheckProducerDefaultScopeTest` AND `SmallRyeHealthProcessor.processSmallRyeHealthConfigValues()` does not reset the `io.smallrye.health.context.propagation` value (the system property is only set if `SmallRyeHealthConfig.contextPropagation==true`). As a result the health check is reinitialized [here](https://github.com/smallrye/smallrye-health/blob/main/implementation/src/main/java/io/smallrye/health/SmallRyeHealthReporter.java#L378).
2. `BeanDefiningAnnotationBuildItem` default scope does not work with a CDI qualifier annotation -> the health check producer method is `@Dependent`.

Problem 2 should be solved in https://github.com/quarkusio/quarkus/pull/27658.

@Ladicek would you care to look at the problem 1?

Ladicek commented:
I believe `SmallRyeHealthProcessor.processSmallRyeHealthConfigValues()` shouldn't really use `SystemPropertyBuildItem`. It should use `RunTimeConfigurationDefaultBuildItem` instead. I'll take a proper look.

Ladicek committed:
Fix SmallRye Health configuration conversion

The SmallRye Health extension needs to take the Quarkus configuration
(`quarkus.smallrye-health.*`) and convert some of it to SmallRye Health
specific configuration properties (`io.smallrye.health.*`). To do that,
the extension used to produce a `SystemPropertyBuildItem`, which results
in changing JVM-global system properties. This may negatively affect
`QuarkusUnitTest` tests, because even though each runs their own Quarkus
application, they all share the same JVM.

With this commit, the extension instead uses `RunTimeConfigurationDefaultBuildItem`,
which is properly scoped per application and doesn't affect JVM-global state.

Ladicek commented:
The test failure seems unrelated to me, but I want to rebase this on [`main` once we merge] #27658 anyway.

mkouba commented:
And pls remove the [TMP debug logging](https://github.com/quarkusio/quarkus/pull/27616/commits/75314c9de095201db11cb14124016fbd2ced190a) commit.

Ladicek commented:
Done.
